src/main/java/com/google/api/gax/grpc/PageStreamingCallable.java

```diff
@@ -60,27 +62,39 @@ public String toString() {
     return String.format("pageStreaming(%s)", callable);
   }
 
-  public ListenableFuture<Iterable<ResourceT>> futureCall(CallContext<RequestT> context) {
-    return Futures.immediateFuture((Iterable<ResourceT>)new StreamingIterable(context));
+  @Override
+  public ListenableFuture<PageAccessor<ResourceT>> futureCall(CallContext<RequestT> context) {
+    PageAccessor<ResourceT> pageAccessor =
+        new PageAccessorImpl<RequestT, ResponseT, ResourceT>(callable, pageDescriptor, context);
+    return Futures.immediateFuture(pageAccessor);
   }
 
-  private class StreamingIterable implements Iterable<ResourceT> {
+  private class PageAccessorImpl<RequestT, ResponseT, ResourceT>
+      implements PageAccessor<ResourceT> {
+    private final FutureCallable<RequestT, ResponseT> callable;
+    private final PageStreamingDescriptor<RequestT, ResponseT, ResourceT> pageDescriptor;
     private final CallContext<RequestT> context;
+    private ResponseT currentPage;
 
-    private StreamingIterable(CallContext<RequestT> context) {
+    private PageAccessorImpl(FutureCallable<RequestT, ResponseT> callable,
+        PageStreamingDescriptor<RequestT, ResponseT, ResourceT> pageDescriptor,
+        CallContext<RequestT> context) {
       this.context = context;
+      this.pageDescriptor = pageDescriptor;
+      this.callable = callable;
+      this.currentPage = null;
     }
 
     @Override
     public Iterator<ResourceT> iterator() {
-      return new StreamingIterator(context.getRequest());
+      return new PageIterator(context.getRequest());
     }
 
-    private class StreamingIterator implements Iterator<ResourceT> {
+    private class PageIterator implements Iterator<ResourceT> {
```

Comment:
It would simplify your code. Taking care of the boilerplate `remove` and should make the next/hasNext handling more simple.